test/inductor/test_compiled_autograd.py

```diff
@@ -1308,6 +1309,57 @@ def fn():
 
         self.check_output_and_recompiles(fn, 1)
 
+    def test_trace_run_with_rng_state(self):
+        def g(xq, xk):
+            # xq: (bs, n_local_heads, seqlen, head_dim)
+            # xk: (bs, n_local_heads, cache_len + seqlen, head_dim)
+            y = F.scaled_dot_product_attention(xq, xk, xk, is_causal=True)
+            z = torch.matmul(y, y)
+            return z
+
+        def h(xq, xk):
+            out = torch.utils.checkpoint.checkpoint(g, xq, xk, use_reentrant=False)
+            return out
+
+        def f():
+            bs = 1
+            n_local_heads = 1
+            seqlen = 2
+            head_dim = 2
+            cache_len = 2
+            xq = torch.ones(
+                (bs, n_local_heads, seqlen, head_dim), requires_grad=True, device="cpu"
+            )
+            xk = torch.ones(
+                (bs, n_local_heads, cache_len + seqlen, head_dim),
+                requires_grad=True,
+                device="cpu",
+            )
+            # `run_with_rng_state` op only shows up when `F.scaled_dot_product_attention`'s forward is compiled
+            out = torch.compile(h, fullgraph=True)(xq, xk)
+            out.sum().backward()
+            return out, xq.grad, xk.grad
+
+        compiler_fn = make_compiler_fn(fullgraph=True)
```

Comment:
@yf225 what would be really helpful here is a walkthrough of what happens with run_with_rng_state. In particular:

1) where does run_with_rng_state show up? (in the forward graph or the backward graph or both?) 
2) what does the dynamo graph captured by compiled autograd look like?
3) do we expect `run_with_rng_state` to stay alive until inductor? If so, the FunctionalTensor kernel below looks incorrect, because it removes run_with_rng_state

Reply:
thanks @zou3519 - added details on the answers to the questions, and updated `run_with_rng_state` functionalization impl so that the op is preserved all the way to inductor.